README
This is Perl module Test::SharedFork.

INSTALLATION

Test::SharedFork installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Test::SharedFork

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Test::SharedFork documentation is available as in POD. So you can do:

    % perldoc Test::SharedFork

to read the documentation online with your favorite pager.

Tokuhiro Matsuno


Changes
Revision history for Perl module Test::SharedFork

{{$NEXT}}

0.35 2015-12-21T23:25:12Z

    - Support Test2

0.34 2015-09-29T14:53:27Z

    - Support upcoming Test::Stream changes

0.33 2015-06-06T03:36:58Z

    - Tell Test::Stream::IPC to poll for results

0.32 2015-06-04T16:14:57Z

    - Update for newer Test::Stream
    - REmove some old Test::Stream support

0.31 2015-04-28T20:25:31Z

    - Fix test that accidentally got commited last release

0.30 2015-04-28T20:19:58Z

    - Minor changes to support newer Test::Stream versions

0.28 2014-07-04T07:37:50Z

    - Fixed minor testing issue on Win32.
      (charsbar++)

0.27 2014-07-03T05:37:00Z

    commit fa8821ff5fe7b44a2461ea58e537ea01221504ed
    Author: Chad Granum <chad.granum@dreamhost.com>
    Date:   Tue Jul 1 15:13:28 2014 -0700

        Fix another issue with Result_Stream branch of TB

0.26 2014-06-27T04:20:47Z

    - Fix support for Result_Stream Test::Builder branch
      (exodist)

0.25 2014-06-24T01:34:10Z

    - Switch to EUMM

0.24 2014-04-07T21:21:56Z

    - Cygwin 64-bit - allow tests to pass
      (rwhitworth)

0.23 2014-03-12T20:12:11Z

    - 06_fail_lineno.t fails on Windows due to path separator #9
      (nanis)

0.22 2014-03-11T00:01:43Z

    - Care the is_passing value.
      https://rt.cpan.org/Public/Bug/Display.html?id=93726
      (Reported by RHOELZ++)

0.21 2012-11-28

    - t/10_subtest.t: added --norc option for more portable testing.

0.20 2012-02-09

    - test_requires Time::HiRes for redhat.
      (berekuk++)

0.19 2011-11-29

    - fixed TB2 support(tokuhirom)

0.18 2011-10-04

    - More fixes for failure locations(hoelzro)

0.17 2011-10-03

    - fixed test level(hoelzro)

0.16 2011-02-10

    - doc tweaks

0.15 2010-09-11

    - added "LIMITATION" section to docs.
      I gave up to support ithreads.

0.14 2010-09-11

    - release!

0.13_01 2010-09-10

    - Test::Builder2 support

0.12 2010-08-17

    - fixed lazy loading issue RT#60426
      (reported by J.)

0.11 2009-12-16

    - write depended perl version. this module requires perl5.8 or later.

0.10 2009-12-14

    - oops

0.09 2009-10-26

    - fixed $Test::Builder::Level(by konbuizm)

0.08 2009-10-18

    - fixed deps for Test::More 0.88(thanks to andk++)

0.07 not released

    - cleanup code.->parent, ->child, ->fork was obsoleted.

0.07_01 2009-10-18

    - win32 support
    - added premitive test cases

0.06 2009-10-15

    - call parent() automatically(reported by kazuho++)

0.05 2009-04-12

    - enhancements for subtests support

0.04 2009-03-25

    - [BUG] previous version does not works with many fork

0.03 2009-03-25

    - added Test::SharedFork->fork()
    - use flock(2) instead of IPC::ShareLite

0.02 2009-03-25

    - ooppppsss. Test::Fork was already used by schwern.
      rename to Test::SharedFork

0.01 2009-03-25 14:06:28

    - original version



lib/Test/Fork.pm
package Test::Fork;
use strict;
use warnings;
our $VERSION = '0.01';
use Test::Builder;
use Test::Fork::Scalar;
use Test::Fork::Array;
use Test::Fork::Store;
use IPC::ShareLite ':lock';
use Storable ();

our $TEST;
sub import {
    $TEST = Test::Builder->new();

    my $store = Test::Fork::Store->new();
    $store->lock_cb(sub {
        $store->{share}->store(Storable::nfreeze(+{
            array => [],
            scalar => 0,
        }));
    }, LOCK_EX);
}

my $CLEANUPME;
sub parent {
    my $store = _setup();
    $CLEANUPME = $store;
}

sub child {
    # And musuka said: 'ラピュタは滅びぬ！何度でもよみがえるさ！'
    # (Quote from 'LAPUTA: Castle in he Sky')
    $TEST->no_ending(1);

    _setup();
}

sub _setup {
    my $store = Test::Fork::Store->new();
    tie $TEST->{Curr_Test}, 'Test::Fork::Scalar', 0, $store;
    tie @{$TEST->{Test_Results}}, 'Test::Fork::Array', $store;

    no strict 'refs';
    no warnings 'redefine';
    for my $name (qw/ok skip todo_skip current_test/) {
        my $cur = *{"Test::Builder::${name}"}{CODE};
        *{"Test::Builder::${name}"} = sub {
            my @args = @_;
            $store->lock_cb(sub {
                $cur->(@args);
            }, LOCK_EX);
        };
    };
    return $store;
}

END {
    if ($CLEANUPME) {
        $CLEANUPME->{share}->destroy(1);
    }
}

1;
__END__

=head1 NAME

Test::Fork - fork test

=head1 SYNOPSIS

    use Test::More tests => 200;
    use Test::Fork;

    my $pid = fork();
    if ($pid == 0) {
        # child
        Test::Fork->child;
        ok 1, "child $_" for 1..100;
    } elsif ($pid) {
        # parent
        Test::Fork->parent;
        ok 1, "parent $_" for 1..100;
        waitpid($pid, 0);
    } else {
        die $!;
    }

=head1 DESCRIPTION

Test::Fork is utility module for Test::Builder.
This module makes forking test!

This module merges test count with parent process & child process.

=head1 METHODS

=over 4

=item parent

call this class method, if you are parent

=item child

call this class method, if you are child.

you can call this method many times(maybe the number of your children).

=back

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom  slkjfd gmail.comE<gt>

yappo

=head1 SEE ALSO

L<Test::TCP>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Test/Fork/Array.pm
package Test::Fork::Array;
use strict;
use warnings;
use base 'Tie::Array';
use Storable ();
use IPC::ShareLite ':lock';

sub _get {
    my $self = shift;

    return $self->{share}->lock_cb(sub {
        $self->{share}->get('array');
    }, LOCK_SH);
}
sub FETCH {
    my ($self, $index) = @_;
    $self->_get()->[$index];
}
sub FETCHSIZE {
    my $self = shift;
    my $ary = $self->_get();
    scalar @$ary;
}

sub TIEARRAY {
    my ($class, $share) = @_;
    my $self = bless { share => $share }, $class;
    $self;
}

sub STORE {
    my ($self, $index, $val) = @_;

    $self->{share}->lock_cb(sub {
        my $share = $self->{share};
        my $cur = $share->get('array');
        $cur->[$index] = $val;
        $share->set(array => $cur);
    }, LOCK_EX);
}

1;


lib/Test/Fork/Store.pm
package Test::Fork::Store;
use strict;
use warnings;
use IPC::ShareLite ':lock';
use Storable ();

sub new {
    my $class = shift;
    my $share = IPC::ShareLite->new(
        -key => 0721, ## no critic
        -create => 'yes',
        -destroy => 'no',
    ) or die $!;
    bless {share => $share, lock => 0}, $class;
}

sub share { shift->{share} }

sub get {
    my ($self, $key) = @_;

    my $ret = $self->lock_cb(sub {
        $self->get_nolock($key);
    }, LOCK_SH);
    return $ret;
}

sub get_nolock {
    my ($self, $key) = @_;
    my $dat = $self->{share}->fetch() or die "cannot get the shared data";
    Storable::thaw($dat)->{$key};
}

sub set {
    my ($self, $key, $val) = @_;

    $self->lock_cb(sub {
        $self->set_nolock($key, $val);
    }, LOCK_EX);
}

sub set_nolock {
    my ($self, $key, $val) = @_;
    my $share = $self->{share};
    my $dat = Storable::thaw($share->fetch());
    $dat->{$key} = $val;
    $share->store(Storable::nfreeze($dat))
}

sub lock {
    my $self = shift;
    $self->{lock}++;
    $self->{share}->lock(@_);
}

sub unlock {
    my $self = shift;
    $self->{lock}--;
    if ($self->{lock} == 0) {
        $self->{share}->unlock();
    }
}

sub lock_cb {
    my ($self, $cb, $type) = @_;
    $type ||= LOCK_EX;

    $self->lock(LOCK_EX);
    my $ret = $cb->();
    $self->unlock();
    $ret;
}

1;


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Test-SharedFork',
    module_name     => 'Test::SharedFork',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


Makefile.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;
use strict;

use ExtUtils::MakeMaker;


my %WriteMakefileArgs = (
    NAME     => 'Test::SharedFork',
    DISTNAME => 'Test-SharedFork',
    VERSION  => '0.34',
    EXE_FILES => [glob('script/*'), glob('bin/*')],
    CONFIGURE_REQUIRES => {
  "ExtUtils::MakeMaker" => 0
}
,
    BUILD_REQUIRES     => {}
,
    TEST_REQUIRES      => {
  "App::Prove" => 0,
  "Test::Builder::Tester" => 0,
  "Test::Requires" => 0,
  "Time::HiRes" => 0
}
,
    PREREQ_PM          => {
  "File::Temp" => 0,
  "Test::Builder" => "0.32",
  "Test::Builder::Module" => 0,
  "Test::More" => "0.88",
  "perl" => "5.008_001"
}
,
);

my $full_prereqs = {
  "App::Prove" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Temp" => 0,
  "Test::Builder" => "0.32",
  "Test::Builder::Module" => 0,
  "Test::Builder::Tester" => 0,
  "Test::More" => "0.88",
  "Test::Requires" => 0,
  "Time::HiRes" => 0,
  "perl" => "5.008_001"
}
;

unless (eval { ExtUtils::MakeMaker->VERSION(6.63_03) }) {
    delete $WriteMakefileArgs{TEST_REQUIRES};
    delete $WriteMakefileArgs{BUILD_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM} = $full_prereqs;
}

unless (eval { ExtUtils::MakeMaker->VERSION(6.52) }) {
    delete $WriteMakefileArgs{CONFIGURE_REQUIRES};
}

unless (eval { ExtUtils::MakeMaker->VERSION(6.57_01) }) {
    use File::Copy;
    copy('META.yml', 'MYMETA.yml')   or warn "META.yml: $!";
    copy('META.json', 'MYMETA.json') or warn "META.json: $!";
}

WriteMakefile(%WriteMakefileArgs);
